src/datacontroller/insertcontent.js

```diff
@@ -0,0 +1,411 @@
+		// <p>x</p>[<img>]<p>z</p> + <p>y</p> => <p>x</p><p>y</p><p>z</p> (no merging, note: after running deletetContents
+		//																	 it's exactly the same case as above)
+		this._mergeSiblingsOf( node, context );
+	}
+
+	/**
+	 * @param {engine.model.Element} node The object element.
+	 * @param {Object} context
+	 */
+	_handleObject( node, context ) {
+		// Try finding it a place in the tree.
+		if ( this._checkAndSplitToAllowedPosition( node ) ) {
+			this._insert( node );
+		}
+		// Try autoparagraphing.
+		else {
+			this._tryAutoparagraphing( node, context );
+		}
+	}
+
+	/**
+	 * @param {engine.model.Node} node The disallowed node which needs to be handled.
+	 * @param {Object} context
+	 */
+	_handleDisallowedNode( node, context ) {
+		// Try inserting its children (strip the parent).
+		if ( node instanceof Element ) {
+			this.handleNodes( node.getChildren(), context );
+		}
+		// Try autoparagraphing.
+		else {
+			this._tryAutoparagraphing( node, context );
+		}
+	}
+
+	/**
+	 * @param {engine.model.Node} node The node to insert.
+	 */
+	_insert( node ) {
+		/* istanbul ignore if */
+		if ( !this._checkIsAllowed( node, [ this.position.parent ] ) ) {
+			// Algorithm's correctness check. We should never end up here but it's good to know that we did.
+			// Note that it would often be a silent issue if we insert node in a place where it's not allowed.
+			log.error(
+				'insertcontent-wrong-position: The node cannot be inserted on the given position.',
+				{ node, position: this.position }
+			);
+
+			return;
+		}
+
+		const livePos = LivePosition.createFromPosition( this.position );
+
+		this.batch.insert( this.position, node );
+
+		this.position = Position.createFromPosition( livePos );
+		livePos.detach();
+
+		// The last inserted object should be selected because we can't put a collapsed selection after it.
+		if ( this._checkIsObject( node ) && !this.schema.check( { name: '$text', inside: [ this.position.parent ] } ) ) {
+			this.nodeToSelect = node;
+		} else {
+			this.nodeToSelect = null;
+		}
+	}
+
+	/**
+	 * @param {engine.model.Node} node The node which could potentially be merged.
+	 * @param {Object} context
+	 */
+	_mergeSiblingsOf( node, context ) {
+		if ( !( node instanceof Element ) ) {
+			return;
+		}
+
+		const mergeLeft = context.isFirst && ( node.previousSibling instanceof Element ) && this.canMergeWith.has( node.previousSibling );
+		const mergeRight = context.isLast && ( node.nextSibling instanceof Element ) && this.canMergeWith.has( node.nextSibling );
+		const mergePosLeft = LivePosition.createBefore( node );
+		const mergePosRight = LivePosition.createAfter( node );
+
+		if ( mergeLeft ) {
+			const position = LivePosition.createFromPosition( this.position );
+
+			this.batch.merge( mergePosLeft );
+
+			this.position = Position.createFromPosition( position );
+			position.detach();
+		}
+
+		if ( mergeRight ) {
+			let position;
+
+			/* istanbul ignore next */
```

Comment:
Why you had `/* istanbul ignore if */` before and `/* istanbul ignore next */ here?